kong/plugins/tcp-log/handler.lua

```diff
@@ -7,6 +7,27 @@ local ngx = ngx
 local timer_at = ngx.timer.at
 
 
+local function get_body_data()
+  local req = ngx.req
+
+  req.read_body()
+  local data = req.get_body_data()
+  if data then
+    return data
+  end
+
+  local file_path = req.get_body_file()
```

Comment:
is this wise? it only gets flushed to disk if it is too big. So this is inherently dangerous, I'd skip this, and tell userss to increase the limit before caching to disk if they need it in the logs anyway

Reply:
![get_raw_data](https://user-images.githubusercontent.com/13045944/83757780-92bd9700-a6a3-11ea-8032-25b4658993e2.png)

`kong.request.get_raw_body()` also includes `get_body_file()`. It is difficult to guarantee that users will set `client_body_buffer_size` to be as high as `client_max_body_size` to force request bodies to be kept in memory. How about tell users to increase the limit of `client_body_buffer_size` and use `kong.request.get_raw_body()` at meanwhile?  When this feature is released in the future, I will add a increase limit suggestion about the `log_body` parameter of the tcp-log plugin at https://docs.konghq.com.

Reply:
I did some experiments before. For **grpc requests**, no matter how the `client_body_buffer_size` is set or how small the request body is, the request body will always be stored in a temporary file. This situation has been encountered in both source installation and run docker kong.

This is not a problem of kong, but a bug of nginx processing HTTP/2 request. I saw nginx fixed it: https://nginx.googlesource.com/nginx/+/2939d5c5fadb3f28a8f8ac85c039bfd007239d3c%5E%21/#F0.

But in openresty-1.15.8.2, the nginx code is still the problematic code before the bug fix.

![openresty_nginx](https://user-images.githubusercontent.com/13045944/83780284-8e9f7280-a6bf-11ea-85da-f58376e34fd4.png)

In this case, if you do not read the temporary file, the grpc request body will not be collected. I will continue to dig this problem tomorrow, and if you know more, please let me know ：）

Reply:
This bug was fixed in [google nginx](https://nginx.googlesource.com/nginx), but it has not been merged by the official nginx so far. At present, this bug still exists in the official nginx code, which causes other projects using the official nginx **will always save the request body to a temporary file when processing grpc requests**.

I just got the [reason](https://github.com/cloudendpoints/esp/issues/395#issuecomment-639431437) from a google developer who proposed this fix, he says that the patch was sent upstream and rejected, see: http://mailman.nginx.org/pipermail/nginx-devel/2018-December/011657.html. The simple summary is that the official knows that this problem exists, but they think that the current fix is not comprehensive enough.


So at present, if we want to get the grpc request body, we do need to read the temporary file.  **One thing that needs to be clarified is that the reading and writing of the file would have happened for grpc requests, not caused by the plugin. If users select to log request body when using the tcp-log plugin, only one more read will be caused.** 

If we use the pdk function `kong.request.get_raw_body()`, will not get the request body for the grpc request since it not support reading from temporary files, and this PR will lose its meaning for gprc part (http part is normal).

I tend to use the previous function implementation to support logging grpc requests. Not everyone will set the switch, but this switch provides a possibility for those who need it. Hope to hear your thoughts.

Reply:
The initial implementation is not good enough. For http requests, it will still cause file reading, ignoring the `client_body_buffer_size` parameter.

Two other options:
1. Use `kong.request.get_raw_body()`, and tell users that only supports logging http request body currently. After nginx improves, it will support logging grpc requests.

2. Add additional processing for grpc requests in `kong.request.get_raw_body()`(this requires a separate PR). `client_body_buffer_size` should be transparent to users. Users should not be concerned about the impact of different protocols on `client_body_buffer_size`, they only need to know that when the request body is less than this value, it can be logged.

Reply:
The first option was chosen.